docs/connections.md

```diff
@@ -0,0 +1,597 @@
+  Specifically, an operator name may contain only characters from the following
+  set: `!$%&*+-/<>?^~|:\,.()[]{}=`. Not every sequence of these characters is a valid operator name, as they can collide with other language constructs.
+  
+Any other names that do not match requirements above, like `HTTP`, `foO` or
+`Make_new` are not allowed and their behavior will not be specified.
+
+In non-pattern contexts, referring to the existing identifier is
+case-insensitive. So `foo` can be referred to either as `foo` or `Foo` with no
+difference.  In pattern context, lower-cased names are used to introduce a
+binding (or a constraint), while upper-cased name will refer to an already bound
+identifier. In short, variable name is allowed to shadow, while constructor 
+always unambiguously refers to an externally visible identifier.
+
+Operator names behave either as:
+* variable names when placed in a prefix position (e.g. `+` in `+ ab`);
+* or as constructor names when placed in an infix position (e.g. `,` in `a,b`).
+
+Number and text literals (like `5` or `"Hello"`) are treated as
+constructor names.
+
+Identifiers can be introduced by a binding (e.g. using assignment or lambda
+argument matching) or when adding type constraints (using type ascriptions). 
+
+## Scope
+Scope is the span in the code which shares the identifiers set of visible,
+available identifiers. Scopes are a span-tree structure, covering the whole program code.
+
+Nested scope is allowed to:
+* access identifiers from the outer scope;
+* shadow identifiers from nested scope with a new binding;
+* introduce new constraints on the identifiers from parent (or own) scopes.
+
+The same identifier may be bound multiple times in the same scope (overloading).
+It is allowed only for method overloads that differ in the type of the `this`
+parameter. This limitation may be relaxed in the future, if proper motivating use-cases are found.
+
+The identifier is always accessible scope-wide, both before and after the line
+introducing it. Some monadic contexts (like IO) can introduce order-dependent
+behavior, however. This is not something that IDE is (or can be) concerned about
+when figuring out connections.
+
+Identifier is bound by using a variable name in the pattern context. Exact behavior depends on the language construction that was to introduce the identifier.
+
+Identifier introduced into a scope is visible only in the scope's subtree
+(lexical scoping). 
+
+Scopes, in the core language, are introduced by:
+* module (file), i.e. the root scope;
+* code blocks, i.e. indented blocks that follow a trailing operator;
+* arrow (lambda) `->` operator for its right-hand side.
+
+Some other constructs seemingly introduce scope (like function definitions) but
+this is because they are desugared into some construct that introduces scope (like lambdas).
+
+TODO: Consider if there are any special rules for signatures on definitions, or
```

Comment:
The identifiers defined in a signature are available in the body of the function, and vice-versa.

Reply:
That's fair. 

And yes, it _is_ tricky, but it's also necessary for the way we're planning on handling dependency in type ascription. We want the ability to be able to define a type function `foo` in the body that can then be used in the signature.